### ISSUE TITLE
[plugin API] new feature for adding custom extensions

### ISSUE THREAD

**Issue by brianlheim** [supercollider/supercollider#5346]
[...] what I would find really nice is a way to add custom extensions to scsynth and make them available to plugins without breaking the existing plugin interface.

For example, the plugin interface could have an API method `void* (*fGetExtension)(const char *)` which allows to obtain a struct or function pointers by name, e.g. "Bela_Context", "Bela_Scope", "Bela_SomeCoolFunction". If the result is NULL, it means the extension is not available. Otherwise, the caller just has to cast the result to the correct type and store it somewhere for later use. I'm sure @scztt also has some thoughts about this :-)

_Originally posted by @Spacechild1 in https://github.com/supercollider/supercollider/pull/5295#discussion_r571510114_

**Comment by Spacechild1** [comment]
Thanks for opening the feature request!

> if someone could provide an example of this "in the wild".

VST2 plugins can use the `audioMasterCanDo` op code to query the host for capabilities, e.g. by calling `hostCallback(plugin, audioMasterCanDo, 0, 0, "reportConnectionChanges", 0.f)`.
There is also the `audioMasterVendorSpecific` callback, which can be used to obtain a struct or function pointer from the host. This is how REAPER implements its (host-side) VST2 extensions (see "Host functions" and "Host context" in https://www.reaper.fm/sdk/vst/vst_ext.php)

The VST3 API has a similar functionality, but they use COM-like interfaces instead of C structs. Each interface is defined by a 16 byte GUID and inherits from `FUnknown`, which has a `queryInterface` method that acts like a C++ dynamic cast. E.g. the plugin can try to cast the host context a desired interface extension; if the cast succeeds, it means the interface is supported and the plugin can use it.

Obviously, the VST3 approach is not appropriate for Supercollider :-) But it shows that there is common need to extend the host application and provide a method for plugins to discover those extension.

Note that the string passed to `fGetExtension` should be reasonably unique. Using fake namespaces like "Bela_Context" or "Bela_SomeFunction" might be sufficient.

---

On the other hand, `fGetExtension` could return an extension *interface table*. I actually think this might be more convenient because you only have to query *once*, save the interface table and use it. The drawback is that it's more tricky to add new methods while staying ABI compatible.

---

Since this feature request would break the API, it might be a good chance to also implement https://github.com/supercollider/supercollider/issues/5347. Actually, some aspects actually overlap!

**Comment by brianlheim** [comment]
> Since this feature request would break the API, it might be a good chance to also implement #5347. Actually, there are some common problems/questions!

yes! sorry, would you like to close this and move this convo there?

i think we are starting to see a few ideas coalesce around API v4 now -- pure C interface, extensibility, maybe even see about cleanup up the scsynth/supernova divide... i like it

**Comment by Spacechild1** [comment]
I think that it makes sense to keep this a seperate feature request, as it builds upon #5347, but not vice versa. Maybe we can first sort out #5347, then we might have a clearer idea on how to exactly implement custom host extensions.

**Comment by Spacechild1** [comment]
> pure C interface, extensibility, maybe even see about cleanup up the scsynth/supernova divide... i like it

These are all important points we should take care of! Maybe an RFC would be a better place after all?

**Comment by brianlheim** [comment]
> I think that it makes sense to keep this a seperate feature request, as it builds upon #5347, but not vice versa.

fine with me!

> Maybe an RFC would be a better place after all?

i don't really see this needing to go through an RFC process -- the requests are all reasonable and seem straightforward to implement, we just need to pick a time to do it. i wouldn't stop anyone from opening an RFC for it, or for requesting one first if it turns out more work is needed. in any case, we have a *lot* to do on this project right now and my view is that this topic should wait until we've caught up on existing conversations and requirements.

**Comment by Spacechild1** [comment]
Thanks for https://github.com/supercollider/supercollider/projects/22! I'll add some more issues there, so we don't lose track.

**Comment by brianlheim** [comment]
no problem, was just about to share that here :)

**Comment by brianlheim** [comment]
feel free to make new issues for anything that isn't already in the tracker
